docs/enhancements/qos.md

```diff
@@ -0,0 +1,407 @@
+Traffic coming from pods on the namespace heading to each destination, Protocol and Port that match the selector will be marked with the corresponding DSCP value.
+The destination can be either CIDR(s) and/or targeting a namespace or one or set of pods of specified namespace.
+Not specifying a PodSelector will apply a rule to all pods in the namespace.
+
+For Marking East/West traffic, ovn-k enables `ovn-encap-tos` option so that DSCP marking from the inner header and apply it to the GENEVE packet.
+Therefore, East/West traffic within the cluster can benefit from this API and not just the North/South traffic.
+
+### Implementation Details/Notes/Constraints
+
+A new API `QoS` under the `k8s.ovn.org/v1` version will be added to `pkg/crd`.
+
+A new controller in OVN-K will watch `QoS`, `Pod` and `Node` objects, which will create the relevant QoS objects and attach them to all of the node local switches in the cluster in OVN - resulting in the necessary flows to be programmed in OVS.
+
+In order to not create an OVN QoS object per pod in the namespace, the controller will also manage AddressSets. 
+For each QoS rule specified in a given `QoS` it'll create an AddressSet, adding only the pods
+whose label matches the PodSelector to it, making sure that new/updated/deleted matching pods are also added/updated/deleted accordingly.
+Rules that do not have a PodSelector will leverage the namespace's AddressSet.
+
+For example, using LGW mode and assuming there's a single node `node1` and the following `QoS` is created:
+
+```yaml
+kind: QoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: qos-external-foo
+  namespace: default
+spec:
+  podSelector:
+    matchLabels:
+      priority: Critical
+  egress:
+  - priority: 100
+    dscp: 59
+    classifier:
+      ports:
+        protocol: "tcp"
+        port: 22
+      to:
+      - cidr: 1.2.3.4/32
+  - priority: 10
+    dscp: 46
+    classifier:
+      to:
+      - cidr: 1.2.3.4/32
+---
+kind: QoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: qos-external-bar
+  namespace: default
+spec:
+  egress:
+  - dscp: 30
+    classifier:
+      to:
+      - cidr: 0.0.0.0/0
+```
+the equivalent of:
+```bash
+ovn-nbctl qos-add node1 to-lport 1000 "ip4.src == <default_ns-qos-external-foo address set> && ip4.dst == 1.2.3.4/32 && tcp && tcp.dst == 22" dscp=59
+ovn-nbctl qos-add node1 to-lport 999 "ip4.src == <default_ns-qos-external-foo address set> && ip4.dst == 1.2.3.4/32" dscp=46
+ovn-nbctl qos-add node1 to-lport 998 "ip4.src == <default_ns address set> && ip4.dst == 0.0.0.0/0" dscp=30
+```
+will be executed.
+The podSelector in `qos-external-foo` QoS object is applied for both first and second QoS rules. So creating a new Pod that matches this selector in the namespace results in its IPs being added to corresponding Address Set.
+The `qos-external-bar` doesn't have podSelector, so QoS rule specified in this object is applied to all pods in the `default` namespace.
+
+```yaml
+kind: QoS
```

Comment:
Shouldn't we call this `NetworkQoS`? Plain `QoS` can be confusing and collide with other kinds. QoS could refer as well to CPU, Memory, GPU, Storage. And since we are doing this only for network I would name it as such. We already have `NetworkPolicy` (and not just `Policy`) so why not `NetworkQoS` :)

Reply:
There is also https://github.com/kubernetes/enhancements/pull/3004/files that could be relevant in the future.

Reply:
yes @VirrageS , `NetworkQoS` seems to be appropriate name for CRD, let's go with it.
Thanks for the KEP link, looks like it's targeting only memory and cpu resource now. as you said we can also see how that can be adapted for networking qos. but i hope this is for future.